### PR TITLE
Test Android with Play provider

### PR DESCRIPTION
**Pull request by yschimke**
Run a single request with the Play provider to confirm HTTP/2 and TLS 1.2

### CODE REVIEW

**Comment by yschimke** [comment]
Test logs

```
01-01 21:01:19.254 18856 18877 I TestRunner: started: testRequestUsesPlayProvider(okhttp.android.test.OkHttpTest)
01-01 21:01:19.257 18856 18877 I TaskRunner: Q10000 scheduled after   0 µs: MockWebServer 56059
01-01 21:01:19.258 18856 18883 I TaskRunner: Q10000 starting              : MockWebServer 56059
01-01 21:01:19.627 18856 18877 I TaskRunner: Q10002 scheduled after   0 µs: OkHttp ConnectionPool
01-01 21:01:19.630 18856 18887 I TaskRunner: Q10002 starting              : OkHttp ConnectionPool
01-01 21:01:19.630 18856 18887 I TaskRunner: Q10002 run again after 300 s : OkHttp ConnectionPool
01-01 21:01:19.631 18856 18886 I TaskRunner: Q10003 scheduled after   0 µs: OkHttp facebook.com applyAndAckSettings
01-01 21:01:19.631 18856 18887 I TaskRunner: Q10002 finished run in 439 µs: OkHttp ConnectionPool
01-01 21:01:19.634 18856 18887 I TaskRunner: Q10003 starting              : OkHttp facebook.com applyAndAckSettings
01-01 21:01:19.635 18856 18887 I TaskRunner: Q10005 scheduled after   0 µs: OkHttp facebook.com onSettings
01-01 21:01:19.636 18856 18888 I TaskRunner: Q10005 starting              : OkHttp facebook.com onSettings
01-01 21:01:19.636 18856 18887 I TaskRunner: Q10003 finished run in   3 ms: OkHttp facebook.com applyAndAckSettings
01-01 21:01:19.636 18856 18888 I TaskRunner: Q10005 finished run in 275 µs: OkHttp facebook.com onSettings
01-01 21:01:19.710  1801  2295 E GnssHAL_GnssInterface: gnssSvStatusCb: a: input svInfo.flags is 8
01-01 21:01:19.710  1801  2295 E GnssHAL_GnssInterface: gnssSvStatusCb: b: input svInfo.flags is 8
01-01 21:01:19.793 18856 18877 I TaskRunner: Q10002 scheduled after   0 µs: OkHttp ConnectionPool
01-01 21:01:19.793 18856 18889 I TaskRunner: Q10002 starting              : OkHttp ConnectionPool
01-01 21:01:19.793 18856 18889 I TaskRunner: Q10002 run again after 300 s : OkHttp ConnectionPool
01-01 21:01:19.794 18856 18889 I TaskRunner: Q10002 finished run in 587 µs: OkHttp ConnectionPool
01-01 21:01:19.794 18856 18877 I TaskRunner: Q10002 scheduled after   0 µs: OkHttp awaitIdle
01-01 21:01:19.794 18856 18889 I TaskRunner: Q10002 starting              : OkHttp awaitIdle
01-01 21:01:19.795 18856 18889 I TaskRunner: Q10002 finished run in 374 µs: OkHttp awaitIdle
01-01 21:01:19.798 18856 18877 I TaskRunner: Q10000 scheduled after   0 µs: OkHttp awaitIdle
01-01 21:01:19.799 18856 18883 I TaskRunner: Q10000 finished run in 541 ms: MockWebServer 56059
01-01 21:01:19.799 18856 18883 I TaskRunner: Q10000 starting              : OkHttp awaitIdle
01-01 21:01:19.800 18856 18883 I TaskRunner: Q10000 finished run in 412 µs: OkHttp awaitIdle
01-01 21:01:19.800 18856 18877 I TestRunner: finished: testRequestUsesPlayProvider(okhttp.android.test.OkHttpTest)
```

File: android-test/src/androidTest/java/okhttp/android/test/OkHttpTest.kt [hunk 1]
```diff
@@ -142,6 +144,36 @@ class OkHttpTest {
...
+    assumeNetwork()
+
+    try {
+      ProviderInstaller.installIfNeeded(InstrumentationRegistry.getTargetContext())
+
+      val request = Request.Builder().url("https://facebook.com/robots.txt").build()
+
+      var socketClass: String? = null
+
+      client = OkHttpClient.Builder().eventListener(object : EventListener() {
```
Comment (yschimke) [inline android-test/src/androidTest/java/okhttp/android/test/OkHttpTest.kt]: Fixing this line up ATM...

File: okhttp/src/main/java/okhttp3/OkHttpClient.kt [hunk 2]
```diff
@@ -736,6 +736,12 @@ open class OkHttpClient internal constructor(
       this.x509TrustManagerOrNull = trustManager
     }
 
+    fun resetSslSocketFactory() = apply {
```
Comment (yschimke) [inline okhttp/src/main/java/okhttp3/OkHttpClient.kt]: TODO: remove

File: android-test/src/androidTest/java/okhttp/android/test/OkHttpTest.kt [hunk 3]
```diff
@@ -118,30 +120,62 @@ class OkHttpTest {
 
       var socketClass: String? = null
 
-      client = OkHttpClient.Builder().eventListener(object : EventListener() {
+      // Need fresh client to reset sslSocketFactoryOrNull
```
Comment (swankjesse) [inline android-test/src/androidTest/java/okhttp/android/test/OkHttpTest.kt]: Good comment.

File: okhttp-testing-support/src/main/java/okhttp3/OkHttpClientTestRule.kt [hunk 4]
```diff
@@ -68,6 +73,7 @@ class OkHttpClientTestRule : TestRule {
   }
 
   @Synchronized private fun addEvent(event: String) {
+    logger.info(event)
```
Comment (swankjesse) [inline okhttp-testing-support/src/main/java/okhttp3/OkHttpClientTestRule.kt]: Is this going to make all our tests incredibly chatty?

Comment (yschimke) [reply okhttp-testing-support/src/main/java/okhttp3/OkHttpClientTestRule.kt]: A bit, but also shows the MockWebserver is currently logging to STDERR in IDE.  Making this optional.  Good catch.

**Comment by swankjesse** [comment]
It’d be neat to buffer them for like 100 ms because then we could log one line for schedule/start/finish for tasks like this:

```
01-01 22:47:24.806 28070 28102 I TaskRunner: Q10002 starting              : OkHttp ConnectionPool
01-01 22:47:24.806 28070 28102 I TaskRunner: Q10002 run again after 300 s : OkHttp ConnectionPool
01-01 22:47:24.806 28070 28102 I TaskRunner: Q10002 finished run in 350 µs: OkHttp ConnectionPool
```

That said, I don’t think we should do this because the complexity far outweighs the benefits.

**Comment by yschimke** [comment]
Or maybe the logging should be one single line (not 3 above) when below 5ms, but 3 when it's anything more?

**Comment by swankjesse** [comment]
Agreed, the problem is to implement that we need to buffer the line for 5 ms.

**Comment by yschimke** [comment]
Easy to implement in the example log handlers.  Wouldn't have any affect on prod code, but we could design what we what dev experience to be like.

**Pull request merged**